ops/model.py

```diff
@@ -797,6 +1001,314 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+      (one they've created or been granted access to).
+
+    All secret events (e.g. SecretChanged) also have a :meth:`SecretEvent.secret` which
+    returns the Secret object associated with that event.
+    """
+    _secret_key_re = re.compile(r"^([a-z](?:-?[a-z0-9]){2,})$")
+    _secret_id_re = re.compile(r"^secret:((?:-?[a-z0-9]){5,})$")
+    _secret_id_prefix = 'secret:'
+
+    def __init__(self, backend: '_ModelBackend',
+                 id: Optional[str] = None,
+                 label: Optional[str] = None,
+                 revision: Optional[int] = None,
+                 is_owned_by_this_unit: bool = False):
+        if not (id or label):
+            raise TypeError('provide an id or a label or both')
+
+        if id is not None:
+            id = self._ensure_id(id)
+
+        if is_owned_by_this_unit and revision is None:
+            # This should only happen in testing, when one instantiates Secret manually.
+            # we set revision to an obviously weird value so that if the behaviour is unexpected,
+            # you know why.
+            revision = -1
+
+        self._backend = backend
+
+        self._id = id
+        self._label = label
+        self._revision = revision
+        self._is_owned_by_this_unit = is_owned_by_this_unit
+
+    @property
+    def id(self) -> Optional[str]:
+        """Unique identifier for this secret in the Juju model.
+
+        The `id` property will be set (not None) if you obtained this secret via one
+        of the following:
+            - `self.[unit|app].add_secret(...)`
+            - `self.model.get_secret(id=<some-id>)`
+            - from a SecretEvent instance.
+
+        If you obtained this secret via `self.model.get_secret(label=<some-label>)`,
+        then the secret ID will be unknown.
+        """
+        return self._id
+
+    @property
+    def label(self) -> Optional[str]:
+        """Label by which this secret is locally known.
+
+        The `label` property will be set (not None) if you obtained this secret via one
+        of the following:
+            - ``self.[unit|app].add_secret(label=<some-label>, ...)``
+            - ``self.model.get_secret(label=<some-label>)``
+            - from a :class:`SecretEvent` instance.
+
+        If you obtained this secret by ``self.model.get_secret(id=<some-id>)``,
+        then the label for this secret will be unknown.
+
+        This label is locally unique. I.e. Juju will check that *this unit* only has one secret
+        with this label at any given time, but any other unit is free to use the same label for
+        one of their secrets (the same secret, or a different one).
+        The label is private: other units can't see it.
+        """
+        return self._label
+
+    @property
+    def revision(self) -> Optional[int]:
+        """The revision number for this secret.
+
+        This will only be visible to the owner of the secret.
+        Secret consumers will get ``None``.
```

Comment:
I wonder if we need 2 classes:
Secret (for owners)
SecretContent (for consumers, returned from secret_get).
It seems like a suboptimal charmer experience to have to "know" what apis work and which don't.
A consumer will always only get the content. The owner always gets the full secret.